pkg/inline/inline.go

```diff
@@ -177,26 +192,66 @@ func (n *Inliner) InlineComponentsInBundle(ctx context.Context, data *bundle.Bun
 
 // readFile from either a local or remote location.
 func (n *Inliner) readFile(ctx context.Context, file bundle.File) ([]byte, error) {
-	url := file.URL
-	if url == "" {
-		return nil, fmt.Errorf("file %v was specified but no file path/url was provided", file)
+	parsed, err := file.ParsedURL()
+	if err != nil {
+		return nil, err
+	}
+	scheme := files.URLScheme(parsed.Scheme)
+	if scheme == files.EmptyScheme {
+		scheme = files.FileScheme
 	}
-	switch {
-	case strings.HasPrefix("gs://", url):
-		return nil, fmt.Errorf("url-type (GCS) not supported; file was %q", url)
-	case strings.HasPrefix("https://", url) || strings.HasPrefix("http://", url):
-		return nil, fmt.Errorf("url-type (HTTP[S]) not supported; file was %q", url)
-	case strings.HasPrefix("file://", url):
-		return n.Reader.ReadFileObj(ctx, file)
-	default:
-		// By default, assume that the user expects to read from the local filesystem.
-		return n.Reader.ReadFileObj(ctx, file)
+	rdr, ok := n.Readers[scheme]
+	if !ok {
+		return nil, fmt.Errorf("could not find file reader for scheme %q for url %q", parsed.Scheme, file.URL)
 	}
+	return rdr.ReadFileObj(ctx, file)
 }
 
-func shortFileUrl(url string) string {
-	if strings.HasPrefix(url, "file://") {
-		url = strings.TrimPrefix(url, "file://")
+// rewriteObjPaths rewrites relative-path'd file-references during component
+// inlining.
+func (n *Inliner) rewriteObjectPaths(ctx context.Context, compFile bundle.File, comp *bundle.ComponentPackage) error {
+	compURL, err := compFile.ParsedURL()
+	if err != nil {
+		return err
+	}
+
+	for i, o := range comp.Spec.ObjectFiles {
+		objURL, err := o.ParsedURL()
+		if err != nil {
+			return err
+		}
+		scheme := files.URLScheme(objURL.Scheme)
+		if scheme == files.EmptyScheme {
+			scheme = files.FileScheme
+		}
+		rw, ok := n.Rewriters[scheme]
+		if !ok {
+			// It's not really an error to not rewrite paths; Most URL schemes won't
```

Comment:
I don't understand enough what's going on here to know, but this smells funny to me.